### DISCUSSION TITLE
Support sending styles in Hyperview fragments

### DISCUSSION THREAD

**Discussion by flochtililoch** [Instawork/hyperview#1074] [Ideas]
Right now, the styles definition is represented under the `<styles>` element, which by spec requires to be an immediate children of the `<screen>` element. Hyperview fragments by spec only return elements that can be appended to a <body> element.

**Proposal:** define a new `<fragment>` element, modeled after the `<screen>` element, that will serve the purpose of nesting content and styles in two separate elements. For example:

<table>
<tr>
<th>Document</th>
<th>Fragment</th>
</tr>
<tr>
<td rowspan="4">

```xml
<doc xmlns="https://hyperview.org/hyperview">
  <screen>
    <styles>
      <style id="primary" color="blue" />
    </styles>
    <body>
      <text style="primary">Hello</text>
    </body>
  </screen>
</doc>
```

</td>
<td>

**Current**

```xml
<text style="secondary">World!</text>
```

</td>
</tr>
<tr>
<td>

**Proposed**

```xml
<doc xmlns="https://hyperview.org/hyperview">
  <fragment>
    <styles>
      <style id="secondary" color="green" />
    </styles>
    <body>
      <text style="secondary">World!</text>
    </body>
  </fragment>
</doc>
```

</td>
</tr>
</table>

We would need for this to update the existing parsing code to change [the validation logic](https://github.com/Instawork/hyperview/blob/e16ffe6f66747478ea8c245485140db5766845e7/src/services/dom/parser.js#L137-L150) while preserving support for the old format, then extract the stylesheet [here](https://github.com/Instawork/hyperview/blob/e16ffe6f66747478ea8c245485140db5766845e7/src/index.js#L293-L294) (like it's done [here](https://github.com/Instawork/hyperview/blob/e16ffe6f66747478ea8c245485140db5766845e7/src/index.js#L192-L193)) and merge it in the existing stylesheet (stored on the React state).

**Comment by adamstep** [comment]
One other approach we could consider is to allow `<styles>` elements anywhere within the XML. This could make it easier for one-off styling sometimes needed for a specific component.

Another question: should `<styles>` elements override global settings, or only apply to the scope (fragment or doc)? I can see benefits of using localized-styles rather than merging the global styles. RN makes this easy to implement by passing multiple stylesheets as the prop.

**Comment by flochtililoch** [comment]
> One other approach we could consider is to allow `<styles>` elements anywhere within the XML. This could make it easier for one-off styling sometimes needed for a specific component.

I thought about this a bit, I think part of one-off styling is really bound to how the system that generates your markup is organized. With the solution we're currently exploring for our backend, we'll have the ability to collocate styles with components, so that'll fill the need here. It's true that when writing the XML manually, it's not the easiest right now to associate a style with an element, but is this a use case you think is worth optimizing for?

Also, as of right now, not every Hyperview components are rendering children elements, so that might also make it unpredictable when this would work - or would be at the expanse of a larger refactor.

> should `<styles>` elements override global settings, or only apply to the scope (fragment or doc)

By global settings, you meant the document, right? As of right now, styles are already scoped to the doc. Could you list some of the benefits you have in mind for scoping the styles to the fragment only?

**Comment by adamstep** [comment]
> I thought about this a bit, I think part of one-off styling is really bound to how the system that generates your markup is organized. With the solution we're currently exploring for our backend, we'll have the ability to collocate styles with components, so that'll fill the need here. It's true that when writing the XML manually, it's not the easiest right now to associate a style with an element, but is this a use case you think is worth optimizing for?

The use case is when we sometimes have a one-off need, for example if a specific element needs a custom width. I'm noticing our design system styles are getting littered with some of these one-off styles, which is bloating the payload for all screens.

Also worth keeping in mind that when adding features to Hyperview core, the open-source users won't have the benefit of our internal system for rolling up all styles. So I could see some users relying on co-located styles with their components, even if we're not doing that ourselves.

I agree it may not be something we need to add immediately, but I want to at least keep the door open for this sort of design in the future.

> > should `<styles>` elements override global settings, or only apply to the scope (fragment or doc)
> 
> By global settings, you meant the document, right? As of right now, styles are already scoped to the doc. Could you list some of the benefits you have in mind for scoping the styles to the fragment only?

This is probably best shown with an example. I believe `<styles>` are scoped to a `<screen>` right now, right?

```xml
<doc xmlns="https://hyperview.org/hyperview" >
  <screen>
    <styles><style id="main" flex="1" backgroundColor="red"  /></styles>
    <body>
      <view style="main" / > <!-- this will be red -->
    </body>
  </screen>

  <screen>
    <styles><style id="main" flex="1" backgroundColor="blue"  /></styles>
    <body>
      <view style="main" / > <!-- this will be blue -->
    </body>
  </screen>
</doc>
```

So there's a precedent that the styles apply only to the context where they are defined, in this case the `<screen>`.
For consistency, it seems like `<fragment>` should follow the same logic. The example below shows the state of the XML after we append a fragment to the main view:

```xml
<doc xmlns="https://hyperview.org/hyperview" >
  <screen>
    <styles>
      <style id="main" flex="1" backgroundColor="red"  />
    </styles>
    <body>
      <view style="main"> <!-- this will be red -->

        <fragment>
          <styles>
            <style id="secondary" color="green" />
          </styles>
          <body>
            <text style="secondary">World!</text>  <!-- this will be green -->
            <view style="main" /> <!-- this will be red, using style from parent screen -->
          </body>
        </fragment>

        <text style="secondary">Hello</text>  <!-- this will not be green -->

      </view>
    </body>
  </screen>
</doc>
```

Now that I've written out this example, I do see some potential issues, which means maybe we don't want to do this.
- there could be duplicate style ids in the XML, if the same fragment is appended multiple times. Of course, ids should be unique.
- schema gets pretty complicated, due to nested `<body>` elements.


One other idea to explore: since the desired effect is to add styles at the screen-level, perhaps this could be done with a behavior. Something like this:
```xml
<view>
  <behavior trigger="load" once="true" action="add-styles">
    <styles>
      <style id="secondary" color="green" />
    </styles> 
  </behavior>
  <text style="secondary">World!</text>
</view>
```

**Comment by flochtililoch** [comment]
Some alternatives:

1. **Inlined styles** 
    style attributes are only applied to their parent element.
```xml
<view href="#">
  <style backgroundColor="red">
    <modifier pressed="true">
      <style backgroundColor="blue" />
    </modifier>
  </style>
  <text>
    <style color="yellow">
      <modifier pressed="true">
        <style backgroundColor="green" />
      </modifier>
    </style>
    Hello world
  </text>
```

2. **`<style>` anywhere**
    styles are combined in a single stylesheet and collisions are removed. cons: can be undeterministic depending on XML parsing.
```xml
<doc xmlns="https://hyperview.org/hyperview">
  <screen>
    <styles>
      <style id="primary" color="blue" />
    </styles>
    <body>
      <text style="primary">Hello</text>
      <style id="primary" color="red" />
      <view>
        <style id="primary" color="green" />
      </view>
    </body>
  </screen>
</doc>
```

3. **namespaced style attributes**
    similar to inlined styles but less verbose syntax (but more XML namespaces)
```
<doc xmlns="https://hyperview.org/hyperview" hv-style="https://hyperview.org/hyperview/styles" hv-style-pressed="https://hyperview.org/hyperview/styles/pressed">
  <screen>
    <body>
      <text style:color="red" style:fontSize="12">
         Hello
        <text hv-style:backgroundColor="blue" hv-style-pressed:backroundColor="red">world</text>
      </text>
    </body>
  </screen>
</doc>
```

**Comment by adamstep** [comment]
Amongst the alternatives listed, I think option 1 for inlined styles is the most straight-forward.

- Like you said, with `<style>` anywhere, we could have some issues especially when styles are loaded through fragments on an existing screen. But this is similar to CSS, so maybe people are used to it?
- The namespaced style attributes idea. may not scale well as we add modifiers. For example, how would we represent styles for a "pressed and selected" modifier? The more modifiers we add, we have an explosion of namespaces to support.

It may help to introduce a `<inline-style>` tag for option 1, to help distinguish inline styles from those with global references.

**Comment by flochtililoch** [comment]
> Like you said, with <style> anywhere, we could have some issues especially when styles are loaded through fragments on an existing screen. But this is similar to CSS, so maybe people are used to it?

I think the general practice in CSS is to avoid rendering `<style>` anywhere as it's blocking the rendering. MDN [suggests](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/style#try_it) `<style>` must be used in the `<head>` portion of the document.

If we were to implement this approach, we probably would need to implement a [render-blocking mechanism](https://html.spec.whatwg.org/multipage/dom.html#render-blocking-mechanism), though maybe we wouldn't need it if we're never planning on supporting chunked responses (i.e. is there any point in doing render blocking if we already have the whole document loaded, so that every styles are available before rendering starts?)

**Comment by flochtililoch** [comment]
For anyone watching this issue, a custom behavior action `add-styles` that injects styles into the main screen's stylesheet has been the solution we went with for the past couple years. It's now available as a community behavior (#990).
